File: Logica_de_Programacao/arquivo.c
#include <stdlib.h>
#include <stdio.h>

void leitura(FILE *fp)
{
    fp = fopen("arquivo.txt", "r");

    if (!feof(fp))
    {
        double num;
        fread(&num, sizeof(double), 1, fp);
        //fscanf(fp, "%lf", &num);
        leitura(fp);

        if (num == (int)num)
        {
            printf("%d", (int)num);
        }
        else
        {
            printf("%lf", num);
        }
    }
    fclose(fp);
}

int main()
{
    FILE *fp = fopen("arquivo.txt", "wb");
    if (fp == NULL)
    {
        printf("O arquivo não pode ser aberto!");
        exit(0);
    }

    //qtd. de numeros(entradas)
    int n = 0;
    scanf("%d", &n);
    int i = 0;
    
    //Lendo dados do teclado e escrevendo-os no arquivo
    while (i < n)
    {
        double d;
        scanf("%lf", &d);
        fprintf(fp, "%lf\n", d);
        //fread(&d, sizeof(double), 1, fp);
        i++;
    }
    leitura(fp);
    fclose(fp);

    //leitura e escrita de tras pra frente dos dados do arquivo

    return 0;
}

File: Logica_de_Programacao/Arquivo.java
import java.io.*;
import java.util.Locale;
import java.math.RoundingMode;
import java.text.DecimalFormat;

public class Arquivo {
    public static String formataDados(String dado) {
        dado = dado.replaceAll(",", ".");
        return dado;
    }

    public static void main(String[] args) {

        try {
            RandomAccessFile raf = new RandomAccessFile("Arquivo.txt", "rw");
            DecimalFormat df = new DecimalFormat();

            // set the file pointer at 0 position
            // metodo seek análogo ao array
            raf.seek(0);

            int n = MyIO.readInt();
            int i = 0;
            // metodo de leitura e escrita sequencial dos dados no arquivo
            while (i < n) {
                float d = MyIO.readFloat();
                // calling this method will advance the file position of the RandomAccessFile by
                // 1
                raf.writeFloat(d);
                // avanço 4 bytes(float) no arquivo
                raf.seek(i * 4);
                raf.readFloat();
                i++;
            }
            /*
             * return the file pointer // long position = raf.getFilePointer(); the position
             * varies according to the number of bytes of the RandomAccessFile
             */
            // função para leitura de tras pra frente
            for (int j = n - 1; j >= 0; j--) {
                raf.seek(j * 4);
                float num = raf.readFloat();
                // printa o dado do ponteiro correspondente
                // inteiro
                if (num % 1 == 0) {
                    // inteiro
                    MyIO.println((int) num);
                } else {
                    df.format(num);
                    String s = num + "";
                    MyIO.println(formataDados(s));
                }

            }

            raf.close();

        } catch (IOException e) {
            e.printStackTrace();
        }
    }

}
